crypto/evp/exchange.c

```diff
@@ -0,0 +1,346 @@
+         * optional.
+         */
+        EVP_KEYEXCH_free(exchange);
+        EVPerr(EVP_F_EVP_KEYEXCH_FROM_DISPATCH,
+               EVP_R_INVALID_PROVIDER_FUNCTIONS);
+        return NULL;
+    }
+
+    return exchange;
+}
+
+void EVP_KEYEXCH_free(EVP_KEYEXCH *exchange)
+{
+    if (exchange != NULL) {
+        int i;
+
+        CRYPTO_DOWN_REF(&exchange->refcnt, &i, exchange->lock);
+        if (i > 0)
+            return;
+        ossl_provider_free(exchange->prov);
+        CRYPTO_THREAD_lock_free(exchange->lock);
+        OPENSSL_free(exchange);
+    }
+}
+
+int EVP_KEYEXCH_up_ref(EVP_KEYEXCH *exchange)
+{
+    int ref = 0;
+
+    CRYPTO_UP_REF(&exchange->refcnt, &ref, exchange->lock);
+    return 1;
+}
+
+EVP_KEYEXCH *EVP_KEYEXCH_fetch(OPENSSL_CTX *ctx, const char *algorithm,
+                               const char *properties)
+{
+    return evp_generic_fetch(ctx, OSSL_OP_KEYEXCH, algorithm, properties,
+                             evp_keyexch_from_dispatch,
+                             (int (*)(void *))EVP_KEYEXCH_up_ref,
+                             (void (*)(void *))EVP_KEYEXCH_free);
+}
+
+int EVP_PKEY_derive_init_ex(EVP_PKEY_CTX *ctx, EVP_KEYEXCH *exchange)
+{
+    int ret;
+    OSSL_PARAM *param = NULL;
+    size_t paramsz = 0;
+
+    ctx->operation = EVP_PKEY_OP_DERIVE;
+
+    if (ctx->engine != NULL)
+        goto legacy;
+
+    if (exchange != NULL) {
+        if (!EVP_KEYEXCH_up_ref(exchange))
+            goto err;
+    } else {
+        int nid = ctx->pkey != NULL ? ctx->pkey->type : ctx->pmeth->pkey_id;
+
+        /*
+         * TODO(3.0): Check for legacy handling. Remove this once all all
+         * algorithms are moved to providers.
+         */
+        if (ctx->pkey != NULL) {
+            switch (ctx->pkey->type) {
+            case EVP_PKEY_DH:
+                break;
+            default:
+                goto legacy;
+            }
+            exchange = EVP_KEYEXCH_fetch(NULL, OBJ_nid2sn(nid), NULL);
+        } else {
+            goto legacy;
+        }
+
+        if (exchange == NULL) {
+            EVPerr(EVP_F_EVP_PKEY_DERIVE_INIT_EX, EVP_R_INITIALIZATION_ERROR);
+            goto err;
+        }
+    }
+
+    if (ctx->exchprovctx != NULL && ctx->exchange != NULL)
+        ctx->exchange->freectx(ctx->exchprovctx);
+    EVP_KEYEXCH_free(ctx->exchange);
+    ctx->exchange = exchange;
+    if (ctx->pkey != NULL) {
+        param = evp_pkey_to_param(ctx->pkey, &paramsz);
+        if (param == NULL) {
+            EVPerr(EVP_F_EVP_PKEY_DERIVE_INIT_EX, EVP_R_INITIALIZATION_ERROR);
+            goto err;
+        }
+    }
+    ctx->exchprovctx = exchange->newctx(ossl_provider_ctx(exchange->prov));
+    if (ctx->exchprovctx == NULL) {
+        OPENSSL_clear_free(param, paramsz);
+        EVPerr(EVP_F_EVP_PKEY_DERIVE_INIT_EX, EVP_R_INITIALIZATION_ERROR);
+        goto err;
+    }
+    ret = exchange->init(ctx->exchprovctx, param);
+    OPENSSL_free(param);
```

Comment:
This should be either secure_free or normal free shouldn't it?

I don't think this needs to be fixed here, it will be resolved by #9305 or #9254 (which ever ends up being selected).  But a TODO might be worthwhile.

Reply:
Yeah, good point. Most of the time it should actually be OPENSSL_secure_clear_free, because we expect the key to have a private element. I've changed it to that for now, and added a TODO.